sdk/Managed/src/Microsoft.WindowsAzure.MobileServices/Table/Sync/Queue/Actions/PushAction.cs

```diff
@@ -132,104 +132,104 @@ internal class PushAction : SyncAction
             // keep taking out operations and executing them until queue is empty or operation finds the bookmark or batch is aborted 
             while (operation != null)
             {
-                bool success = await this.ExecuteOperationAsync(operation, batch);
-
-                if (batch.AbortReason.HasValue)
+                using (await this.OperationQueue.LockItemAsync(operation.ItemId, this.CancellationToken))
                 {
-                    break;
-                }
+                    bool success = await this.ExecuteOperationAsync(operation, batch);
 
-                if (success)
-                {
-                    // we successfuly executed an operation so remove it from queue
-                    await this.OperationQueue.DeleteAsync(operation.Id, operation.Version);
-                }
+                    if (batch.AbortReason.HasValue)
+                    {
+                        break;
+                    }
 
-                // get next operation
-                operation = await this.OperationQueue.PeekAsync(operation.Sequence, this.tableKind, this.tableNames);
+                    if (success)
+                    {
+                        // we successfuly executed an operation so remove it from queue
+                        await this.OperationQueue.DeleteAsync(operation.Id, operation.Version);
+                    }
+
+                    // get next operation
+                    operation = await this.OperationQueue.PeekAsync(operation.Sequence, this.tableKind, this.tableNames);
```

Comment:
This has already been merged, but in any case, it's not absolutely necessary that the peek be under the lock, but it's a step in the right direction, as this "batch" isn't really a batch at all since operations keep creeping into the push action while in-flight.